### ISSUE TITLE
tactics: cardinality constraints to CNF

### ISSUE THREAD

**Issue by teobaluta** [Z3Prover/z3#1730]
I am using z3py to convert cardinality constraints into CNF. The tactics I used are `t = Then('simplify', 'card2bv', 'simplify', 'bit-blast', 'tseitin-cnf')`. My goal has 100 constraints over around 800 variables. The conversion takes around 48 minutes on an Intel Xeon E5 CPU and the formula is really big (> 10 million  clauses and 2 mil variables).

1. Is the strategy that I used the most efficient one for this kind of constraints in terms of compactness or speed?

2. Does z3 implement something like sequential counters from Sinz, 2005 [1]?

[1] http://www.carstensinz.de/papers/CP-2005.pdf

**Comment by NikolajBjorner** [comment]
Z3 implements a handful of strategies for converting PB and cardinality constraints. 
The default for PB is to convert to bit-vectors and add them pairwise using log(k)+1 bits, where k is the upper bound, so overflows are checked on each addition. It is also possible to set the sat.pb.solver parameter to other encodings (run z3 -pm:sat for the names of these settings). 

For cardinality constraints it uses sorting circuits and I tried a couple of other encodings that are derivatives of Sinz. The blowup is significant and sounds like a bug. The overhead should be n log n where n is probably the combined size of your problem (at most 80K). It doesn't share sub-circuits or anything more sophisticated. Given the numbers you quote, I suspect there is something fishy in the conversion that should be addressed. It is probably possible to scale your example down to something smaller that still exhibits undue overhead. In this case, could you share it?

Note that you can solve cardinality and PB constraints with the newest SAT core in Z3 without blasting them. You set the parameters sat.cardinality.solver to "true", and sat.pb.solver to "solver" (if you have more than cardinality constraints).

**Comment by teobaluta** [comment]
Thanks for the prompt response!

I want to use z3 as a model converter though. So I just want to get the SAT encoding for the constraints. Does the python API expose any of the options you mentioned?

Sure, I will try to scale it and share it.

**Comment by NikolajBjorner** [comment]
All features are available from python. You can set parameters on tactics or globally

**Comment by NikolajBjorner** [comment]
Indeed, the size seems to quadruple with every doubling of parameter.

   from z3 import *
   import sys

   for sz in range(4, 20):
     xs = [Bool("x%d" % i) for i in range(sz)]
     fml = AtMost(xs +[sz/2])
     g = Goal()
     g.add(fml)
     t = Tactic("card2bv")
     r = t(g)
     g = r[0]
     print(sz, len(g))
     sys.stdout.flush()

**Comment by NikolajBjorner** [comment]
You can try different encoders now.
The name of the parameter is "sat.cardinlaity.encoding", and relevant values are "grouped" (default), "circuit" (corresponding to the log(k)+1 bits), and "unate" (correspoding to unate encoding).
I also revised some settings to get smaller sorting circuits. However, the improvements are not staggering and I don't see a way to produce smaller encodings. The circuit encoding size is significant.

**Comment by NikolajBjorner** [comment]
Here is a sample script

    from z3 import *
    import sys


    set_option("sat.cardinality.encoding", "circuit")
    set_option("sat.cardinality.encoding", "sorting")
    # blows up: set_option("sat.cardinality.encoding", "unate")

    def mk_card(sz):
      xs = [Bool("x%d" % i) for i in range(sz)]
      fml = AtMost(xs +[sz/2])
      g = Goal()
      g.add(fml)
      # t = Then("card2bv", "tseitin-cnf")
      # t = Tactic("card2bv")
      t = Then("card2bv", "simplify", "tseitin-cnf")
      r = t(g)
      g = r[0]
      print(sz, len(g))
      #print(g)
      sys.stdout.flush()
     

    mk_card(4) 
    mk_card(8) 
    mk_card(17) 
    mk_card(33)
    mk_card(65)
    mk_card(129)
    mk_card(257)

   #for sz in range(4, 20):
    #  mk_card(sz)
